### PR TITLE
Advanced register substitution

### PR DESCRIPTION
**Pull request by hbs**
Added support for replacing variables in [ ... ] STORE construct. Fixed register allocation when registers are already used in the macro

### CODE REVIEW

File: warp10/src/main/java/io/warp10/script/functions/ASREGS.java [hunk 1]
```diff
@@ -49,17 +50,131 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
...
+          // Fetch what precedes the CSTORE
+          Object symbol = statements.get(i - 1);
+          if (symbol instanceof Long) {
+            inuse.set(((Long) symbol).intValue());
+          } else if (!(symbol instanceof String)) {
+            // We encountered a CSTORE with a non string and non long param,
+            // we cannot determine what is being updated, so no replacement
+            // can occur
+            abort = true;
+          }
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/functions/ASREGS.java]: Can be merged with `LOAD` block as the code is the same.

File: warp10/src/main/java/io/warp10/script/functions/ASREGS.java [hunk 2]
```diff
@@ -108,37 +216,62 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
...
-            statements.set(i, regfuncs[regno]);
-          }
         } else if (i > 0 && statements.get(i) instanceof STORE) {
           Object symbol = statements.get(i - 1);
-          if (!(symbol instanceof String)) {
+          if (symbol instanceof String) {
+            Integer regno = varregs.get(symbol.toString());
+            if (null != regno) {
+              statements.set(i - 1, NOOP);
+              statements.set(i, regfuncs[regno+regidx]);
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/functions/ASREGS.java]: Missing spaces around `+`.

File: warp10/src/main/java/io/warp10/script/functions/ASREGS.java [hunk 3]
```diff
@@ -108,37 +216,62 @@ public Object apply(WarpScriptStack stack) throws WarpScriptException {
...
-            statements.set(i, regfuncs[regno+regidx]);
-          }        
         } else if (i > 0 && statements.get(i) instanceof CSTORE) {
           Object symbol = statements.get(i - 1);
-          if (!(symbol instanceof String)) {
+          if (symbol instanceof String) {
+            Integer regno = varregs.get(symbol.toString());
+            if (null != regno) {
+              statements.set(i - 1, NOOP);
+              statements.set(i, regfuncs[regno+regidx+regidx]);
```
Comment (ftence) [inline warp10/src/main/java/io/warp10/script/functions/ASREGS.java]: Missing spaces around `+`.

**Comment by hbs** [comment]
PTAL

**Pull request merged**